geode-core/src/main/java/org/apache/geode/internal/cache/map/RegionMapDestroy.java

```diff
@@ -372,6 +372,8 @@ private void retryRemoveWithTombstone() {
           } finally {
             removeEntryOrLeaveTombstone();
           }
+        } else if (regionEntry == null && !newRegionEntry.isTombstone()) {
```

Comment:
I think it would be better to change it so the only the we don't do if we are evicting is handleEntryNotFound(). When evicting we do not want an exception if the entry is already gone.
So I would change line 369 to just be an "else" and then put a "if (!isEviction)" on handleEntryNotFound (even better might be to change handleEntryNotFound to test isEviction and return since this is the only place it is called).
Then you can get rid of your new code on 375 and 376.
I do think it is probably correct to call removeEntryOrLeaveTombstone for an eviction just to make sure it follows the normal rules of scheduling a tombstone when concurrency checks are enabled. This method is already a noop if evicting without concurrency checks.

Reply:
I moved a bit fast here with the first solution, as it appears we fail to pass the unit test evictDestroyOfExistingTombstoneWithConcurrencyChecksReturnsFalse. Suggesting that we don't properly handle the case where we're in an eviction and already set the region entry as tombstone. If we break up the check on isEviction and newRegionEntry.isTombstone we cause more unit tests to fail.

With the current issue, we're trying to catch the specific case where we have concurrency checks enabled, we're in an eviction, and we've created a new region entry marked as tombstone. I've restructured a some to avoid a clause with a lone call to removeEntryOrLeaveTombstone that addresses this issue and passes unit tests.